# Описание проекта


Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.


Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку

**Описание данных:**

- children — количество детей в семье
 - days_employed — общий трудовой стаж в днях
 - dob_years — возраст клиента в годах
 - education — уровень образования клиента
 - education_id — идентификатор уровня образования
 - family_status — семейное положение
 - family_status_id — идентификатор семейного положения
 - gender — пол клиента
 - income_type — тип занятости
 - debt — имел ли задолженность по возврату кредитов
 - total_income — ежемесячный доход
 - purpose — цель получения кредита

# Исследование надежности заемщиков


## Изучение общей информаци о данных

In [4]:
try:
    data = pd.read_csv("/datasets/data.csv")# прочитайте csv-файл
except FileNotFoundError:
    data = pd.read_csv("data.csv")

In [5]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [7]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

 В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [8]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [9]:
data['days_employed'] = data['days_employed'].abs()

In [10]:
#медианное значение трудового стажа days_employed в днях
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. 

In [11]:
#уникальные значений столбца children
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть два аномальных значения. Удалим строки

In [13]:
data = data[(data['children'] != -1) & (data['children'] != 20)]
#проверка
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

In [15]:
#Заполниие пропусков в столбце `days_employed` медианными значениями
#по каждого типа занятости `income_type`
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()
    
#проверка
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [16]:
#замена тип данных в столбце total_income на целочисленный с помощью метода astype()
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [17]:
data['education'] = data['education'].str.lower()

In [18]:
#количество дубликатов
data.duplicated().sum()

71

In [19]:
#удаление дубликатов
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [22]:
#уникальные цели кредита
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

 Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

# Исследование данных

## Сущетвует ли зависимость между количеством детей и возвратом кредита в срок?

In [26]:
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

In [28]:
#  функция для формирования сводной таблицы, с переименованием столбцов и добавления расчета в процентах

def pvt(index, columns):
    data_pvt = data.pivot_table(index = index,
                            columns = columns,
                            values = 'total_income',
                            aggfunc = 'count',fill_value=0,
                            margins = True)
    
    data_pvt = data_pvt.rename(columns={0:"without_dept",1:"in_dept"})

    data_pvt['without_dept_%'] = (data_pvt ['without_dept']/data_pvt['All']*100).round(2)
    data_pvt['in_dept_%'] = (data_pvt ['in_dept']/data_pvt['All']*100).round(2)
    return data_pvt

    
data_pivot_status = pvt('children','debt')
display(data_pivot_status)

debt      without_dept  in_dept    All  without_dept_%  in_dept_%
children                                                         
0                13028     1063  14091           92.46       7.54
1                 4364      444   4808           90.77       9.23
2                 1858      194   2052           90.55       9.45
3                  303       27    330           91.82       8.18
4                   37        4     41           90.24       9.76
5                    9        0      9          100.00       0.00
All              19599     1732  21331           91.88       8.12

**Вывод:**
 Наиболее ответсвенные,согласно отчету, бездетные клиенты. Процент выплаты кредита в срок составляет  92.46%, в противовес им можно выделить клиентов с двумя детьми (90,55%) и с одним ребенком (90,77%).Тем самым можно опровергнуть стереотип, что бездетные клиенты менее ответственны.

## Существует ли зависимость между семейным положением и возвратом кредита в срок?

In [29]:
# запуск функции pvt (сводная таблица) с заданными переменными 'family_status','debt'
data_pivot_income = pvt('family_status','debt')
display(data_pivot_income)

debt                   without_dept  in_dept    All  without_dept_%  in_dept_%
family_status                                                                 
Не женат / не замужем          2523      273   2796           90.24       9.76
в разводе                      1105       84   1189           92.94       7.06
вдовец / вдова                  888       63    951           93.38       6.62
гражданский брак               3749      385   4134           90.69       9.31
женат / замужем               11334      927  12261           92.44       7.56
All                           19599     1732  21331           91.88       8.12

**Вывод:** 
На основе полученных результатов можно сделать следующие выводы: самые безответственные клиенты - не женатые/не замужние,процент невозврата кредита составляет 9,7% и клиенты в гражданском браке - 9,3% невозрата.
Вдовцы прилежно выполняют свои обязанности и являются "передовиками" (6,6% невозврата), но являются самой малочисленной группой. Самая многочисленная группа -клиенты в браке, процент невозрата сотавляет 7,5%, что в целом говорит о их надежности.

## Существует ли зависимость между уровнем дохода и возвратом кредита в срок?

In [32]:
data_pivot_income = pvt('total_income_category','debt')
display(data_pivot_income)

debt                   without_dept  in_dept    All  without_dept_%  in_dept_%
total_income_category                                                         
A                                23        2     25           92.00       8.00
B                              4660      354   5014           92.94       7.06
C                             14568     1353  15921           91.50       8.50
D                               328       21    349           93.98       6.02
E                                20        2     22           90.91       9.09
All                           19599     1732  21331           91.88       8.12

**Вывод:**
Согласно отчету можно выделить две крупные категории - B и C.
Категория C ( доход от 50 000 до 200 000) самая многочисленная, процент закрытия долга в срок составляет 91,50%. Категорию B (с доходом от 200 000 до 1 000 000) можно выделить как более надежных клиентов, процент закрытия долга 92,94%.

## Как разные цели кредита влияют на его возврат в срок?

In [34]:
data_pivot_income = pvt('purpose_category','debt')
display(data_pivot_income)

debt                      without_dept  in_dept    All  without_dept_%  \
purpose_category                                                         
операции с автомобилем            3879      400   4279           90.65   
операции с недвижимостью          9971      780  10751           92.74   
получение образования             3619      369   3988           90.75   
проведение свадьбы                2130      183   2313           92.09   
All                              19599     1732  21331           91.88   

debt                      in_dept_%  
purpose_category                     
операции с автомобилем         9.35  
операции с недвижимостью       7.26  
получение образования          9.25  
проведение свадьбы             7.91  
All                            8.12

In [37]:
display(data.groupby('purpose_category')['debt'].agg(['count', 'sum', 'mean'])
        .sort_values(by = 'mean', ascending = False))


count  sum      mean
purpose_category                              
операции с автомобилем     4279  400  0.093480
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118
операции с недвижимостью  10751  780  0.072551

**Вывод:**
Согласно отчету, вероятность невыплаты кредита самая низкая у кредитов на  "операции с недвижимостью" - 7,2% и на "проведение свадьбы" - 7,9%. Наибольший процент невыплаты у кредитов на " операции с автомобилем", где процент невозврата составляет 9,3%

# Общий вывод

**Итоги исследования:**

В ходе анализа были совершены следующие действия над данными:
- удалены пропуски в столбцах days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход) и заменены на медианные значения по каждому типу занятости;
- выявлены и удалены строки с повторяющими значениям, а также строки с аномальными значениями, что искажают данные общей статистики;
- добавлены стобцы с категоризацией клиентов по уровню дохода (категории A-E,где А- это клиенты с самым высоким уровнем дохода, Е- самым низким) и также добавлена категоризация по цели получения кредита (все цели приведены в единую форму и разбиты на 4 основные группы: операции с автомобилем,получение образования, проведение свадьбы, операции с недвижимостью)

Все эти действия позволили сделать наиболее точный анализ и сформировать следующие выводы:
- наиболее доходный сегмент клиентов, которые показывают высокий процент возврата кредита в срок и являются наиболее надежными - клиенты без детей(процент закрытия долга в срок 92,46%), находящиеся в браке ( 92,44%), с уровнем дохода от 200 000 до 1000 000(92,94%), совершающие операции с недвижимостью ( 92,74%)

- стоит отметить категорию клиентов, с наибольшим риском. В данном случае можно выделить клиентов с двумя детьмя (процент невозрата кредита в срок 9,45%),не женатые/не замужние (9,76%), с уровнем дохода от 50 000 до 200 000(8,5%), совершающие операции с автомобилем (9,35%)

- Самой многочисленной группой являются клиенты состоящие в браке, без детей, с доходом от 50 000 до 200 000, совершающие операции с недвижимостьюю Это целевая аудитория, на которую стоит обратить внимание и разработать комплекс программ.
